### DISCUSSION TITLE
Question about text-input

### DISCUSSION THREAD

**Discussion by maron5466** [dialogic-godot/dialogic#2238] [Q&A] [answered]
Hi!
![4ad5341176d5e2dcc5e77751c61ff58b01d750f2](https://github.com/dialogic-godot/dialogic/assets/170271202/662aa8ce-079f-4b26-a3b7-ed03e8e6076c)
I am trying to have the player insert a number in the text_input in Dialogic 2 so that the player can receive the money he is asking. The sommaprestito variable is not recognized as a number, and I get the error E 0:00:12:0901 subsystem_expression.gd:36 @ execute_string(): Invalid named index ‘sommaprestito’ for base type Object
<C++ Error> Condition “p_show_error” is true. Returning: Variant()
<C++ Source> core/math/expression.cpp:1507 @ execute()
subsystem_expression.gd:36 @ execute_string()
subsystem_expression.gd:44 @ execute_condition()
event_condition.gd:27 @ _execute()               But I am not sure how to use it, I think maybe text_input only takes strings. How can I have the player insert a float instead?

**Comment by Jowan-Spooner** [comment]
Actually this looks like the right way to do it. TextInput event will always return a string, but you should be able to convert it to a float just like you tried to. 
Which dialogic version are you using? I just did a quick test, and it seemed to work for me (on the main branch)

**Comment by maron5466** [comment]
I am not really sure, I know it is Dialogic 2, but I downloaded it a few months ago. the error is still E 0:00:23:0382   subsystem_expression.gd:36 @ execute_string(): Invalid named index 'sommaprestito' for base type Object
  <C++ Error>    Condition "p_show_error" is true. Returning: Variant()
  <C++ Source>   core/math/expression.cpp:1507 @ execute()
  <Stack Trace>  subsystem_expression.gd:36 @ execute_string()
                 subsystem_expression.gd:44 @ execute_condition()
                 event_condition.gd:27 @ _execute()
I jus got the elif "It is not a real number, I cannot lend you anything." even if before it showed me a sommaprestito = 10, after I entered 10. If enter letters, it shows sommaprestito = 0. I add here a text editor screenshot too.
<img width="668" alt="Screenshot 2024-05-19 at 23 40 45" src="https://github.com/dialogic-godot/dialogic/assets/170271202/2d7f1ee5-63ad-45f2-a22d-4a9d0bf9ab61">

**Reply by Jowan-Spooner** [reply]
Yeah tbh, I can't really help you with older versions of dialogic. I'd recommend giving the main branch a go (preferable after having commited or stashed all your current changes ;).

**Reply by maron5466** [reply]
I am sorry, what do you mean with the main branch?

**Reply by CakeVR** [reply]
The `main` branch means the Dialogic GitHub repository's branch named `main`. It contains the latest changes.

If you need further help: https://docs.dialogic.pro/getting-started.html#-how-to-install-the-latest-wip-version-

Always back up your changes, preferably with git version control.

**Reply by maron5466** [reply, marked as answer]
Thank you very much, I finally solved the problem: the problem was that elif was not complete. Now it is 	elif {sommaprestito} == 0:   and everything works.   Still, there is something I do not understand. Whatever I do, Godot refuses to let me change the global.nave1.gold. It recognizes its value (global is the name of a singleton I am using), but it refuses to change it, even if I tried set {global.nave1.gold} += {sommaprestito}. Is there something I am missing in Dialogic 2 and the use of singleton variables?

**Reply by Jowan-Spooner** [reply]
@maron5466 what is this nave1 value? If it's a dictionary, dialogic only allows accessing dictionaries with the dictionary["key"] format not dictionary.key right now.
Otherwise Idk.

**Reply by maron5466** [reply]
I am sorry. nave1 is an object from the class "nave" I declared in the singleton global, and gold is one of its properties.

**Reply by Jowan-Spooner** [reply]
Then it should work. Damn. Not sure if you are still using an older version. I'd try updating to [Dialogic 2.0 alpha 14](https://github.com/dialogic-godot/dialogic/releases/tag/2.0-alpha-14). If it's not working in that version, feel free to open a bug report issue.

**Reply by maron5466** [reply]
In the end, now it works. I updated to alpha 14. Can I ask a last question? After the text_input, and my variable global.sommatotale, I wanted to active a function called global.chiedere_prestito(banchiere), which needs the variable sommaprestito and the variable banchiere I was planning to set. But I am not really sure how I should call the function in dialogic 2. How should I do?

**Reply by Jowan-Spooner** [reply]
Hey @maron5466, I completely missed this.
You should be able to call the method with the Call event:
`do global.chiedere_prestito(10)`
In that method you can access the variables like usual: `Dialogic.VAR.sommaprestito`

**Reply by maron5466** [reply]
Thank you very much. I need to ask a last question: using the choices menu, I am trying to have some of the choices (not all of them) appear only if a certain condition happens (if global.nave1.number > 0, for example). All of my attempts using the if condition have failed.

**Reply by Jowan-Spooner** [reply]
The correct syntax should be:
```gdscript
- Some Choice [if global.variable > 10] [else="hide" ] 
# from alpha 15 onwards:
- Some Choice | [if global.variable > 10] [else="hide" ]
```
If that doesn't work please start a separate thread on discord for example or open a dedicated issue.

**Reply by maron5466** [reply]
Thank you very much for the answer. I am sorry if I ask again, but if I desire to do something like - Some Choice [if global.variable > 10 and if global.variable2 ==false] [else="hide" ]   , what should I do? It is mportnat for my project to have multiple if conditions

**Reply by Jowan-Spooner** [reply]
Well that works fine as well, but when combining conditions there should be no additional `if` after the `and`/`or`. So your example should say
`- Some Choice [if global.variable > 10 and global.variable2 == false] [else="hide" ]`

**Reply by maron5466** [reply]
Thank you very much. I am sorry if I ask here what I hope are the last questions: 1) where can I find a description of what I should use as a coding syntax in Dialogic 2? 2) For another choice, I would like that it was based on if any(c.destination == "string 1" for c in global.lista_quest). But I have no clues on what I should write, I just need this so that the choice is activated if any of the quests have a destination which coincides with the current city.

**Reply by Jowan-Spooner** [reply]
Hey @maron5466, the documentation is still somewhat lacking. But you can find an overview of the text syntax here: https://docs.dialogic.pro/timeline-text-syntax.html. 

For expressions check this page: https://docs.dialogic.pro/variables.html#conditions-text-syntax. For more complex checks I suggest having a function on an autoload that returns true or false and then calling that method in the expression, e.g. 

`- Repair the thing | [if {Autoload.can_repair("Thing")}]`
```gdscript
# on the Autoload
func can_repair(what:String) -> bool:
     match what:
          "Thing":
               if Dialogic.VAR.RepairSkill > 10 and Dialogic.VAR.Inventory.Stuff > 1:
                    return true
      return false 
```
